co-simulation/fed/mosaic-carma-cloud/src/main/java/org/eclipse/mosaic/fed/carmacloud/ambassador/CarmaCloudMessageAmbassador.java

```diff
@@ -0,0 +1,224 @@
+	 *                            CarmaCloudMessageAmbassador.
+	 */
+	public CarmaCloudMessageAmbassador(AmbassadorParameter ambassadorParameter)
+	{
+		super(ambassadorParameter);
+		try // load configuration file
+		{
+			carmaCloudConfiguration = new ObjectInstantiation<>(CarmaCloudConfiguration.class, log)
+				.readFile(ambassadorParameter.configuration);
+		}
+		catch (InstantiationException e)
+		{
+			log.error("Configuration object could not be instantiated: ", e);
+		}
+
+		log.info("The update interval of CARMA Cloud message ambassador is {}.", carmaCloudConfiguration.updateInterval);
+		
+		if (carmaCloudConfiguration.updateInterval <= 0L)
+		{
+			throw new RuntimeException("Invalid update interval for CARMA Cloud message ambassador, should be > 0.");
+		}
+		log.info("CARMA Cloud message ambassador is generated.");
+	}
+
+
+	/**
+	 * This method is called to tell the federate the start time and the end time.
+	 *
+	 * @param startTime Start time of the simulation run in nanoseconds.
+	 * @param endTime   End time of the simulation run in nanoseconds.
+	 * @throws InternalFederateException Exception is thrown if an error is occurred
+	 *                                   while execute of a federate.
+	 */
+	@Override
+	public void initialize(long startTime, long endTime)
+		throws InternalFederateException
+	{
+		super.initialize(startTime, endTime);
+		currentSimulationTime = startTime;
+
+		carmaCloudRegistrationReceiver = new CarmaCloudRegistrationReceiver();
+		carmaCloudRegistrationReceiver.init();
+		registrationRxBackgroundThread = new Thread(carmaCloudRegistrationReceiver);
+		registrationRxBackgroundThread.start();
+
+		try
+		{
+			rti.requestAdvanceTime(currentSimulationTime, 0, (byte) 1);
+		}
+		catch (IllegalValueException e)
+		{
+			log.error("Error during advanceTime request", e);
+			throw new InternalFederateException(e);
+		}
+	}
+
+
+	/**
+	 * This method processes the interactions.
+	 *
+	 * @param interaction The interaction that can be processed.
+	 * @throws InternalFederateException Exception is thrown if an error is occurred
+	 *                                   while execute of a federate.
+	 */
+	@Override
+	public void processInteraction(Interaction interaction)
+		throws InternalFederateException
+	{
+		String type = interaction.getTypeId();
+		long interactionTime = interaction.getTime();
+		log.trace("Process interaction with type '{}' at time: {}", type, interactionTime);
+	}
+
+
+	/**
+	 * This method is called by the AbstractFederateAmbassador when the RTI grants a
+	 * time advance to the federate.Any unprocessed interactions are forwarded to
+ the federate using the processInteraction method before this call is made.
+	 *
+	 * @param time The timestamp (in nanoseconds) indicating the time to which the federate can
+	 *             advance its local time.
+	 * @throws InternalFederateException
+	 */
+	@Override
+	public synchronized void processTimeAdvanceGrant(long time)
```

Comment:
I know we are not great about this currently in our ambassadors but we should be setting our current simulation time to the time provided here. Otherwise by incrementing incorrectly we could introduce an offset.

Reply:
Do you mean we should do `currentSimulationTime = time` rather than `currentSimulationTime += carmaCloudConfiguration.updateInterval * TIME.MILLI_SECOND;`